classes/tdx-signed-fit-image.inc

```diff
@@ -21,3 +21,14 @@ FIT_HASH_ALG ?= "sha256"
 FIT_SIGN_ALG ?= "rsa2048"
 FIT_SIGN_NUMBITS ?= "2048"
 FIT_SIGN_INDIVIDUAL = "0"
+
+# TODO: Get rid of this block of assignments involving :k3 once fixes to meta-ti are upstreamed.
+# NOTE: A side effect of these assignments is that users will need to use :forcevariable to set them.
+#
+# stronger assignments to override settings in meta-ti
+UBOOT_SIGN_KEYDIR:k3 ?= "${TOPDIR}/keys/fit"
+UBOOT_SIGN_KEYNAME:k3 ?= "dev"
+UBOOT_SIGN_IMG_KEYNAME:k3 ?= "dev2"
+FIT_HASH_ALG:k3 ?= "sha256"
+FIT_SIGN_ALG:k3 ?= "rsa2048"
+FIT_SIGN_NUMBITS:k3 ?= "2048"
```

Comment:
Can you explain better what was the build error you had without setting the FIT_SIGN_NUMBITS variable? Do you have the logs to share? I don't remember having errors when building in my environment without this variable. I just want to understand why you have the error in your environment and I don't.

Reply:
IIRC, I got the exact same error that you reported in our internal ticket (TOR-3556), namely:

```
|   Timestamp:    Thu Jan  1 00:00:00 1970
| Signature written to '/home/sprado/projects/toradex/build/torizon/scarthgap/build-verdin-am62/tmp/work/verdin_am62-tdx-linux/u-boot-toradex-ti/2024.04+git/build/fitImage-linux', node '/configurations/conf-verd
in-am62_spidev_overlay.dtbo/signature-1'                                                                                                                                                                           
| Public key written to 'u-boot.dtb', node '/signature/key-custMpk'
| Signature check bad (error 1)                                                                          
| Verifying Hash Integrity for node 'conf-ti_k3-am625-verdin-nonwifi-dahlia.dtb'... sha512,rsa4096:custMpk-
|  error!                                                                                                
| Verification failed for '(null)' hash node in 'conf-ti_k3-am625-verdin-nonwifi-dahlia.dtb' config node 
| Failed to verify required signature 'key-custMpk'                                                      
| WARNING: exit code 1 from a shell command.                                                             
ERROR: Task (/home/sprado/projects/toradex/build/torizon/scarthgap/build-verdin-am62/conf/../../layers/meta-toradex-ti/recipes-bsp/u-boot/u-boot-toradex-ti_2024.04.bb:do_uboot_assemble_fitimage) failed with exit
 code '1'
```

And I ran a couple of build tests where I found the error happened when `FIT_SIGN_ALG` and `FIT_SIGN_NUMBITS` were not aligned.

Anyways, I'll re-run some tests to confirm and report back here...

Reply:
Ok,

1. Removing any assignments to `FIT_HASH_ALG`, `FIT_SIGN_ALG` and `FIT_SIGN_NUMBITS` from our layer (the normal ones and the ones with the `k3` override) the resolved values of these variables becomes:

```
FIT_HASH_ALG = "sha512"
FIT_SIGN_ALG = "rsa4096"
FIT_SIGN_NUMBITS = <not-set> # when baking virtual/bootloader
FIT_SIGN_NUMBITS = "2048"    # when baking virtual/kernel
```

In this case, the build fails with pretty much the same error (except for the name of the key, whose assignment I kept in the layer):

```
| Signature written to '/workdir/build-verdin-am62-sec/tmp/work/verdin_am62-tdx-linux/u-boot-toradex-ti/2024.04+git/build/fitImage-linux', node '/configurations/conf-verdin-am62_spidev_overlay.dtbo/signature-1'
| Public key written to 'u-boot.dtb', node '/signature/key-dev'
| Signature check bad (error 1)
| Verifying Hash Integrity for node 'conf-ti_k3-am625-verdin-nonwifi-dahlia.dtb'... sha512,rsa4096:dev-
|  error!
| Verification failed for '(null)' hash node in 'conf-ti_k3-am625-verdin-nonwifi-dahlia.dtb' config node
| Failed to verify required signature 'key-dev'
| WARNING: /workdir/build-verdin-am62-sec/tmp/work/verdin_am62-tdx-linux/u-boot-toradex-ti/2024.04+git/temp/run.do_uboot_assemble_fitimage.1310264:213 exit 1 from 'uboot-fit_check_sign -k "u-boot.dtb" -f /workdir/build-verdin-am62-sec/tmp/work/verdin_am62-tdx-linux/u-boot-toradex-ti/2024.04+git/build/fitImage-linux'
| WARNING: Backtrace (BB generated script):
|       #1: concat_dtb, /workdir/build-verdin-am62-sec/tmp/work/verdin_am62-tdx-linux/u-boot-toradex-ti/2024.04+git/temp/run.do_uboot_assemble_fitimage.1310264, line 213
|       #2: uboot_assemble_fitimage_helper, /workdir/build-verdin-am62-sec/tmp/work/verdin_am62-tdx-linux/u-boot-toradex-ti/2024.04+git/temp/run.do_uboot_assemble_fitimage.1310264, line 186
|       #3: do_uboot_assemble_fitimage, /workdir/build-verdin-am62-sec/tmp/work/verdin_am62-tdx-linux/u-boot-toradex-ti/2024.04+git/temp/run.do_uboot_assemble_fitimage.1310264, line 172
|       #4: main, /workdir/build-verdin-am62-sec/tmp/work/verdin_am62-tdx-linux/u-boot-toradex-ti/2024.04+git/temp/run.do_uboot_assemble_fitimage.1310264, line 391
ERROR: Task (/workdir/build-verdin-am62-sec/conf/../../layers/meta-toradex-ti/recipes-bsp/u-boot/u-boot-toradex-ti_2024.04.bb:do_uboot_assemble_fitimage) failed with exit code '1'
```

2. Repeating the same test but adding only the assignment `FIT_SIGN_NUMBITS="4096"` (in order to match the setting `FIT_SIGN_ALG="rsa4096"` coming from meta-ti) the build succeeds and the image boots normally.

From this I think we can conclude that the issue is just the mismatch between `FIT_SIGN_ALG` and `FIT_SIGN_NUMBITS` caused by the setting of only `FIT_SIGN_ALG` in `meta-ti` which I believe is the thing we need to fix upstream.

Reply:
Gotcha. Thanks for testing!